pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -182,3 +313,99 @@ func (w *HashAggPartialWorker) shuffleIntermData(finalConcurrency int) {
 		w.outputChs[i] <- &w.partialResultsMap[i]
 	}
 }
+
+func (w *HashAggPartialWorker) prepareForSpill() {
+	if !w.isSpillPrepared {
+		w.tmpChksForSpill = make([]*chunk.Chunk, spilledPartitionNum)
+		w.spilledChunksIO = make([]*chunk.DataInDiskByChunks, spilledPartitionNum)
+		for i := 0; i < spilledPartitionNum; i++ {
+			w.tmpChksForSpill[i] = w.getNewSpillChunkFunc()
+			w.spilledChunksIO[i] = chunk.NewDataInDiskByChunks(w.spillChunkFieldTypes)
+			if w.spillHelper.diskTracker != nil {
+				w.spilledChunksIO[i].GetDiskTracker().AttachTo(w.spillHelper.diskTracker)
+			}
+		}
+		w.isSpillPrepared = true
+	}
+}
+
+func (w *HashAggPartialWorker) spillDataToDisk() error {
+	err := w.spillDataToDiskImpl()
+	if err == nil {
+		err = generateRandomError()
+	}
+	return err
+}
+
+func (w *HashAggPartialWorker) spillDataToDiskImpl() error {
+	defer func() {
+		if r := recover(); r != nil {
+			recoveryHashAgg(w.globalOutputCh, r)
+		}
+	}()
+	if len(w.partialResultsMap) == 0 {
+		return nil
+	}
+
+	w.prepareForSpill()
+	for _, partialResultsMap := range w.partialResultsMap {
+		for key, partialResults := range partialResultsMap {
+			partitionNum := int(murmur3.Sum32([]byte(key))) % spilledPartitionNum
+
+			// Spill data when tmp chunk is full
+			if w.tmpChksForSpill[partitionNum].IsFull() {
+				err := w.spilledChunksIO[partitionNum].Add(w.tmpChksForSpill[partitionNum])
+				if err != nil {
+					return err
+				}
+				w.tmpChksForSpill[partitionNum].Reset()
+			}
+
+			// Serialize agg meta data to the tmp chunk
+			for i, aggFunc := range w.aggFuncs {
+				aggFunc.SerializePartialResult(partialResults[i], w.tmpChksForSpill[partitionNum], w.spillSerializeHelpers)
+			}
+
+			// Append key
+			w.tmpChksForSpill[partitionNum].AppendString(len(w.aggFuncs), key)
+		}
+	}
+
+	// Clear the partialResultsMap
+	w.partialResultsMap = make([]aggfuncs.AggPartialResultMapper, len(w.partialResultsMap))
```

Comment:
> Can we set `w.partialResultsMap = nil` directly since it would not be used further

I think `w.partialResultsMap` will be used again as more data will continue to be received after the spill done.